### PR TITLE
Validator Page: New Items

### PR DESCRIPTION
**Pull request by TopETH**

### CODE REVIEW

File: src/components/validators/SubnetsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  color: ${theme.palette.error.main};
+`;
+
+export type SubnetsTableProps = {
+	subnets: PaginatedResource<Subnet>;
+	registrations?: bigint[];
+	validatorPermits?: bigint[];
+	initialSortOrder?: string;
+	onSortChange?: (orderBy: SubnetsOrder) => void;
+	initialSort?: string;
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:29]:
```suggestion
```
We don't need these props, do we?

File: src/components/validators/SubnetsTable.tsx [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	},
+	createdAt: {
+		[SortDirection.ASC]: "CREATED_AT_ASC",
+		[SortDirection.DESC]: "CREATED_AT_DESC",
+	},
+	emission: {
+		[SortDirection.ASC]: "ID_ASC",
+		[SortDirection.DESC]: "ID_DESC",
+	},
+};
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:47]:
```suggestion
```

File: src/components/validators/SubnetsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+function SubnetsTable(props: SubnetsTableProps) {
+	const { subnets, registrations, validatorPermits } = props;
+
+	const emissions = useSubnetEmissions();
+
+	const { initialSort, onSortChange } = props;
+	const [sort, setSort] = useState<SortOrder<string>>();
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:55]:
```suggestion
```

File: src/components/validators/SubnetsTable.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+			emissions === undefined
+		)
+			return subnets.data;
+		if (sort?.property !== "emission") return subnets.data;
+		return subnets.data.sort((left: Subnet, right: Subnet) => {
+			if (sort?.direction === SortDirection.ASC)
+				return emissions[left.netUid] - emissions[right.netUid];
+			return emissions[right.netUid] - emissions[left.netUid];
+		});
+	}, [subnets, emissions, sort]);
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:71]:
```suggestion
```

File: src/components/validators/SubnetsTable.tsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+				direction: SortDirection.DESC,
+			});
+		}
+	};
+
+	useEffect(() => {
+		if (!onSortChange || !sort?.property || sort.direction === undefined)
+			return;
+		onSortChange((orderMappings as any)[sort.property][sort.direction]);
+	}, [JSON.stringify(sort)]);
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:108]:
```suggestion
```

File: src/components/validators/SubnetsTable.tsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	return (
+		<ItemsTable
+			data={sortedSubnets}
+			loading={subnets.loading}
+			notFound={subnets.notFound}
+			notFoundMessage="No subnets found"
+			error={subnets.error}
+			data-test="subnets-table"
+			sort={sort}
+			onSortChange={handleSortChange}
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:119]:
```suggestion
```

File: src/components/validators/SubnetsTable.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		>
+			<SubnetsTableAttribute
+				label="ID"
+				sortable
+				render={(subnet) => <>{subnet.netUid}</>}
+				sortProperty="netUid"
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:125]:
```suggestion
```

File: src/components/validators/SubnetsTable.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+			notFound={subnets.notFound}
+			notFoundMessage="No subnets found"
+			error={subnets.error}
+			data-test="subnets-table"
+			sort={sort}
+			onSortChange={handleSortChange}
+		>
+			<SubnetsTableAttribute
+				label="ID"
+				sortable
```
Comment (cuteolaf) [inline src/components/validators/SubnetsTable.tsx:123]:
```suggestion
```

File: src/screens/validators.tsx [hunk 9]
```diff
@@ -167,6 +170,11 @@ export const ValidatorPage = () => {
 		}
 	}, [tab]);
 
+	const subnetsInitialOrder: SubnetsOrder = "NET_UID_ASC";
+	const [subnetSort, setSubnetSort] =
+    useState<SubnetsOrder>(subnetsInitialOrder);
+	const subnets = useSubnets(undefined, subnetSort);
```
Comment (cuteolaf) [inline src/screens/validators.tsx:176]:
```suggestion
	const subnets = useSubnets(undefined);
```

File: src/screens/validators.tsx [hunk 10]
```diff
@@ -264,6 +276,21 @@ export const ValidatorPage = () => {
...
+						label="Subnets"
+						count={subnets.pagination.totalCount}
+						loading={subnets.loading}
+						error={subnets.error}
+						value="subnets"
+					>
+						<SubnetsTable
+							subnets={subnets}
+							onSortChange={(sortKey: SubnetsOrder) => setSubnetSort(sortKey)}
+							initialSort={subnetsInitialOrder}
```
Comment (cuteolaf) [inline src/screens/validators.tsx:289]:
```suggestion
```

File: src/model/validator.ts [hunk 11]
```diff
@@ -12,6 +12,10 @@ export type Validator = {
 	totalDailyReturn: bigint;
 	validatorStake: bigint;
 	validatorReturn: bigint;
+	registrations: string;
+	parsedRegistrations?: bigint[];
```
Comment (cuteolaf) [inline src/model/validator.ts:16]:
```suggestion
```

File: src/model/validator.ts [hunk 12]
```diff
@@ -12,6 +12,10 @@ export type Validator = {
...
+	validatorPermits: string;
+	parsedValidatorPermits?: bigint[];
```
Comment (cuteolaf) [inline src/model/validator.ts:18]:
```suggestion
```

File: src/model/validator.ts
Comment (cuteolaf) [inline src/model/validator.ts:1]: please define a new type with parsed registrations & validator permits

**Review by cuteolaf — CHANGES_REQUESTED** [review]

File: src/model/validator.ts (hunk 12 above)
Comment (TopETH) [reply src/model/validator.ts:18]: Why do you want to remove this? this variable is used.

**Review by cuteolaf — APPROVED** [review]

**Pull request merged**